### PR TITLE
Amazon Auth V4 Support

### PR DESCRIPTION
**Pull request by alanedwardes**
Includes changes to the Amazon S3 support to use the `AWSSDK`, and `PUT` instead of `POST`.

Supports the new [V4 authentication](http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html) mechanism to enable the use of the new Frankfurt region, and future S3 regions.

Fixes https://github.com/ShareX/ShareX/issues/561.

### CODE REVIEW

File: ShareX/Forms/MainForm.cs [hunk 1]
```diff
@@ -46,7 +46,6 @@ public partial class MainForm : HotkeyForm
         private bool forceClose, firstUpdateCheck = true;
         private UploadInfoManager uim;
         private ToolStripDropDownItem tsmiImageFileUploaders, tsmiTrayImageFileUploaders, tsmiTextFileUploaders, tsmiTrayTextFileUploaders;
-        private System.Threading.Timer updateTimer;
```
Comment (Jaex) [inline ShareX/Forms/MainForm.cs]: Why you removed auto updating?

Comment (alanedwardes) [inline ShareX/Forms/MainForm.cs]: This variable wasn't in use

On 20 Apr 2015 8:59 am, Jaex <notifications@github.com> wrote:

In ShareX/Forms/MainForm.cs<https://github.com/ShareX/ShareX/pull/636#discussion_r28669517>:

> @@ -46,7 +46,6 @@ public partial class MainForm : HotkeyForm
>          private bool forceClose, firstUpdateCheck = true;
>          private UploadInfoManager uim;
>          private ToolStripDropDownItem tsmiImageFileUploaders, tsmiTrayImageFileUploaders, tsmiTextFileUploaders, tsmiTrayTextFileUploaders;
> -        private System.Threading.Timer updateTimer;


Why you removed auto updating?

Comment (Jaex) [inline ShareX/Forms/MainForm.cs]: It wasn't in use in Debug but in use in Release.

**Comment by Jaex** [comment]
Is DreamObjects support works with this?

**Comment by alanedwardes** [comment]
No, as I said in the original issue, DreamObjects does not support the AWS V4 authentication scheme.

On 20 Apr 2015 9:04 am, Jaex <notifications@github.com> wrote:

Is DreamObjects support works with this?

File: ShareX/Forms/MainForm.cs (hunk 1 above)
Comment (alanedwardes) [inline ShareX/Forms/MainForm.cs]: OK, will restore later.

**Comment by McoreD** [comment]
Need to inform the guys at DreamHost. Users of http://wiki.dreamhost.com/How_to_Use_ShareX_with_DreamObjects aren't gonna be happy.

**Comment by Jaex** [comment]
I'm not really sure is supporting one region more important than removing DreamObjects support :(

**Comment by alanedwardes** [comment]
I think it would be better to poke them to update their authentication schemes, if they’re going for a like for like match of the S3 API, they need to implement the new stuff.

**Comment by alanedwardes** [comment]
It’s entirely up to you – merge it or don’t.

Bear in mind this futureproofs it, any other regions Amazon releases (and they will release more) will use V4 only, so support will gradually drop off. I’m unsure if they will deprecate the V2 authentication scheme for existing regions though, I haven’t read anything that indicates they will.

**Comment by alanedwardes** [comment]
Have tweeted them, will see what they say: https://twitter.com/AlanEdwardes/status/590113749194924032

When I get home I’ll check to see if it’s possible to drop the AWS SDK back down to auth V2 for arbitrary endpoints. I don’t really want to have the old code hanging around just for DO… we could make a separate file uploader for DreamObjects though, that would make more sense since the APIs are diverging. Leave it with me, will update the PR.

**Comment by alanedwardes** [comment]
Closing this while I work on it.

**Comment by alanedwardes** [comment]
The SDK does allow dropping down to the old auth scheme. Re-opened, includes DreamObjects support again :dancers:

**Comment by Jaex** [comment]
Thanks.

**Pull request merged**